### PR TITLE
Image classifier changes

### PR DESCRIPTION
**Pull request by mattdangerw**
Some more image classification changes suggested by @fchollet 

- Move image classifier implementations to the base class.
- Simplify the `ImageConverter`. One class, rescaling is optional, resizing is optional.

Moving the image classifier code to the base class mean you can do thing like below.

```python
keras_hub.models.ImageClassifier(
    backbone=custom_backbone,
    preprocessor=custom_preprocessor,
    num_classes=2,
)
```

### CODE REVIEW

**Review by divyashreepathihalli — COMMENTED** [review]
Should we still continue to have backbone specific classifier classes?
can we have one ImageClassifierTask and save presets like `resnet_50_image_classifier_imagenet`,  `densenet_image_classifier_imagenet`?

**Comment by mattdangerw** [comment]
> Should we still continue to have backbone specific classifier classes?

Right now, we still have a subclass per architecture. Most vision models don't need anything there. But a backbone with different pooling (e.g. a ViTImageClassifier), could change things in the subclass.

We could revisit this. But it would be a larger change to get loading across keras/transformers/timm right, and be a little more difficult to explain. (TextClassifier has subclasses, CausalLM has subclasses, but ImageClassifier does not?)

> can we have one ImageClassifierTask and save presets like resnet_50_image_classifier_imagenet, densenet_image_classifier_imagenet?

I'm confused at this part. Our resnet presets already include a classification head.

```python
classifier = keras_hub.ImageClassifier.from_preset("resnet_50_imagenet")
classifier.predict(image) # imagenet labels
```

What is different about what you are suggesting? Why are you including classifier in the preset name? What would resnet_50_imagenet be if not a classifier?

**Comment by mattdangerw** [comment]
@divyashreepathihalli at a high-level, ResNet is still our reference implementation of what a classification model should look like. Most other models are still some level of incomplete (missing weights, missing preprocessing, missing conversion, missing tests, etc).

File: keras_hub/src/models/image_classifier.py [hunk 1]
```diff
@@ -16,10 +16,145 @@ class ImageClassifier(Task):
...
+        "resnet_50_imagenet"
+    )
+    classifier.predict(images)
+    ```
+
+    Call `fit()` on a single batch.
+    ```python
+    # Load preset and train
+    images = np.random.randint(0, 256, size=(2, 224, 224, 3))
+    labels = [0, 3]
```
Comment (fchollet) [inline keras_hub/src/models/image_classifier.py:54]: Q: are the code snippets tested? Does this work with Python lists (as opposed to np arrays or tensors)?

File: keras_hub/src/models/image_classifier.py [hunk 2]
```diff
@@ -16,10 +16,145 @@ class ImageClassifier(Task):
...
 
-    All `ImageClassifier` tasks include a `from_preset()` constructor which can be
-    used to load a pre-trained config and weights.
+    Args:
+        backbone: A `keras_hub.models.Backbone` instance or a `keras.Model`.
+        num_classes: int. The number of classes to predict.
+        preprocessor: `None`, a `keras_hub.models.Preprocessor` instance,
+            a `keras.Layer` instance, or a callable. If `None` no preprocessing
+            will be applied to the inputs.
+        pooling: `"avg"` or `"max"`. The type of pooling to apply on backbone
```
Comment (fchollet) [inline keras_hub/src/models/image_classifier.py:25]: We should add `"flatten"` as an option. It is used by VGG models. It's not exactly pooling but that's ok.

File: keras_hub/src/models/image_classifier.py [hunk 3]
```diff
@@ -16,10 +16,145 @@ class ImageClassifier(Task):
...
+            activation=activation,
+            dtype=head_dtype,
+            name="predictions",
+        )
+
+        # === Functional Model ===
+        inputs = self.backbone.input
+        x = self.backbone(inputs)
+        x = self.pooler(x)
+        outputs = self.output_dense(x)
```
Comment (fchollet) [inline keras_hub/src/models/image_classifier.py:134]: Some common models aren't expressible in this format because they require more than one dense layer. It is also extremely common to need dropout.

I think we should add an argument for dropout (applied before the last dense layer). We may also want to make the classification head fully configurable to be able to express more presets with the same class. Alternatively we could leave this to the e.g. `VGGImageClassifier` subclass to implement, in order to keep things simple in the base class.

File: keras_hub/src/models/vgg/vgg_image_classifier.py [hunk 4]
```diff
@@ -1,110 +1,8 @@
...
-        self.output_dense = keras.layers.Dense(
-            num_classes,
-            activation=activation,
-            name="predictions",
-        )
-
-        # === Functional Model ===
-        inputs = self.backbone.input
-        x = self.backbone(inputs)
-        outputs = self.output_dense(x)
```
Comment (fchollet) [inline keras_hub/src/models/vgg/vgg_image_classifier.py:88]: Note -- The real VGG16 classifier has a flatten layer and 2 dense layers.

**Review by fchollet — COMMENTED** [review]
The code looks good!

I think we should disable resizing by default on all architectures that are able to process variable image sizes (only specify a fixed image size in the preset if the model actually requires it). This means reuploading the presets.

File: keras_hub/src/models/vgg/vgg_image_classifier.py (hunk 4 above)
Comment (mattdangerw) [reply keras_hub/src/models/vgg/vgg_image_classifier.py:88]: Got it, I can add this back.

File: keras_hub/src/models/image_classifier.py (hunk 3 above)
Comment (mattdangerw) [reply keras_hub/src/models/image_classifier.py:134]: Adding dropout sounds good.

I'd say let's leave more advanced customization to the subclasses like `VGGImageClassifier`. More consistent with text models.

If you really want custom as an end users, take the backbone and do whatever with it. We already have good coverage for that journey I think.

File: keras_hub/src/models/image_classifier.py (hunk 1 above)
Comment (mattdangerw) [reply keras_hub/src/models/image_classifier.py:54]: I'll double check. We don't have automatic testing for code blocks, because in general they would be quite expensive (downloading 7b parameter models in some cases). Would be great to add a nightly job for this someday.

Comment (mattdangerw) [reply keras_hub/src/models/image_classifier.py:54]: This works.

File: keras_hub/src/models/vgg/vgg_image_classifier.py (hunk 4 above)
Comment (mattdangerw) [reply keras_hub/src/models/vgg/vgg_image_classifier.py:88]: This is done! But one bit of weirdness, in the vgg subclass, I have to skip chaining to the ImageClassifier constructor to avoid defining two functional models. I chain directly to the Task base class instead.

There's a lot of different ways we could fix that up if we wanted to, but might be better done as a follow up.

File: keras_hub/src/layers/preprocessing/image_converter.py [hunk 5]
```diff
@@ -1,46 +1,151 @@
...
 from keras_hub.src.utils.preset_utils import get_preset_loader
 from keras_hub.src.utils.preset_utils import get_preset_saver
 from keras_hub.src.utils.python_utils import classproperty
+from keras_hub.src.utils.tensor_utils import preprocessing_function
 
 
 @keras_hub_export("keras_hub.layers.ImageConverter")
 class ImageConverter(PreprocessingLayer):
-    """Convert raw image for models that support image input.
+    """Preprocess raw images into model ready inputs.
```
Comment (divyashreepathihalli) [inline keras_hub/src/layers/preprocessing/image_converter.py:22]: The args section is missing in the docstring for this layer.

File: keras_hub/src/layers/preprocessing/image_converter_test.py [hunk 6]
```diff
@@ -15,6 +16,50 @@
 
 
 class ImageConverterTest(TestCase):
+    def test_resize_simple(self):
```
Comment (divyashreepathihalli) [inline keras_hub/src/layers/preprocessing/image_converter_test.py:19]: should the self.run_layer_test be added for this layer?

**Review by divyashreepathihalli — APPROVED** [review]
Thanks Matt! left 2 NIT comments

File: keras_hub/src/layers/preprocessing/image_converter.py (hunk 5 above)
Comment (mattdangerw) [reply keras_hub/src/layers/preprocessing/image_converter.py:22]: done!

File: keras_hub/src/layers/preprocessing/image_converter_test.py (hunk 6 above)
Comment (divyashreepathihalli) [reply keras_hub/src/layers/preprocessing/image_converter_test.py:19]: sounds good!
